File: apps/cli-go/internal/protocol/messages.go
package protocol

import (
	"encoding/json"
)

// Envelope is used for initial type discrimination when parsing messages.
type Envelope struct {
	Type string `json:"type"`
}

// AuthMessage is sent by the client to authenticate a tunnel session.
type AuthMessage struct {
	Type      string `json:"type"`
	Subdomain string `json:"subdomain"`
	TTL       int    `json:"ttl,omitempty"`
	SessionID string `json:"sessionId,omitempty"`
}

// AuthAckMessage is sent by the server to acknowledge a successful authentication.
type AuthAckMessage struct {
	Type             string `json:"type"`
	Subdomain        string `json:"subdomain"`
	URL              string `json:"url"`
	TTL              int    `json:"ttl"`
	RemainingTTL     int    `json:"remainingTtl"`
	SessionID        string `json:"sessionId"`
	MaxBodySizeBytes int    `json:"maxBodySizeBytes"`
}

// HttpRequestMessage represents an incoming HTTP request forwarded through the tunnel.
type HttpRequestMessage struct {
	Type    string            `json:"type"`
	ID      string            `json:"id"`
	Method  string            `json:"method"`
	Path    string            `json:"path"`
	Headers map[string]string `json:"headers"`
	HasBody bool              `json:"hasBody"`
}

// HttpResponseMetaMessage contains the response metadata sent back through the tunnel.
type HttpResponseMetaMessage struct {
	Type    string            `json:"type"`
	ID      string            `json:"id"`
	Status  int               `json:"status"`
	Headers map[string]string `json:"headers"`
	HasBody bool              `json:"hasBody"`
}

// HttpBodyChunkMessage signals a body chunk transfer for a given request.
type HttpBodyChunkMessage struct {
	Type string `json:"type"`
	ID   string `json:"id"`
	Done bool   `json:"done"`
}

// HttpRequestEndMessage signals the end of an HTTP request body.
type HttpRequestEndMessage struct {
	Type string `json:"type"`
	ID   string `json:"id"`
}

// HttpResponseEndMessage signals the end of an HTTP response body.
type HttpResponseEndMessage struct {
	Type string `json:"type"`
	ID   string `json:"id"`
}

// PingMsg is a keep-alive ping message.
type PingMsg struct {
	Type string `json:"type"`
}

// PongMsg is a keep-alive pong response message.
type PongMsg struct {
	Type string `json:"type"`
}

// ErrorMessage is sent by the server to indicate an error.
type ErrorMessage struct {
	Type      string `json:"type"`
	Message   string `json:"message"`
	RequestID string `json:"requestId,omitempty"`
	Status    int    `json:"status,omitempty"`
}

// ---- WebSocket relay messages ----

// WsUpgradeMessage is sent by the server to ask the CLI to open a local WS connection.
type WsUpgradeMessage struct {
	Type     string            `json:"type"`
	StreamID string            `json:"streamId"`
	Path     string            `json:"path"`
	Headers  map[string]string `json:"headers"`
}

// WsUpgradeAckMessage is sent by the CLI to confirm or reject a WS upgrade.
type WsUpgradeAckMessage struct {
	Type     string `json:"type"`
	StreamID string `json:"streamId"`
	OK       bool   `json:"ok"`
	Error    string `json:"error,omitempty"`
}

// WsFrameMessage signals a WebSocket frame relay. The actual payload follows
// as a binary frame using the standard binary encoding (streamId prefix + body).
type WsFrameMessage struct {
	Type      string `json:"type"`
	StreamID  string `json:"streamId"`
	FrameType string `json:"frameType"` // "text" or "binary"
}

// WsCloseMessage signals that one side of a relayed WebSocket has closed.
type WsCloseMessage struct {
	Type     string `json:"type"`
	StreamID string `json:"streamId"`
	Code     int    `json:"code"`
	Reason   string `json:"reason"`
}

// ParseTextMessage parses a raw JSON message into its concrete message struct.
// It returns (nil, nil) for unknown message types.
// It returns (nil, error) for malformed JSON.
func ParseTextMessage(raw []byte) (any, error) {
	var env Envelope
	if err := json.Unmarshal(raw, &env); err != nil {
		return nil, err
	}

	switch env.Type {
	case "auth":
		var msg AuthMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "auth-ack":
		var msg AuthAckMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "http-request":
		var msg HttpRequestMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "http-response-meta":
		var msg HttpResponseMetaMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "http-body-chunk":
		var msg HttpBodyChunkMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "http-request-end":
		var msg HttpRequestEndMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "http-response-end":
		var msg HttpResponseEndMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "ping":
		var msg PingMsg
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "pong":
		var msg PongMsg
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "error":
		var msg ErrorMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "ws-upgrade":
		var msg WsUpgradeMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "ws-upgrade-ack":
		var msg WsUpgradeAckMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "ws-frame":
		var msg WsFrameMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	case "ws-close":
		var msg WsCloseMessage
		if err := json.Unmarshal(raw, &msg); err != nil {
			return nil, err
		}
		return &msg, nil

	default:
		return nil, nil
	}
}

// IsTunnelMessage checks whether the given data looks like a tunnel protocol
// message by verifying it contains a "type" field.
func IsTunnelMessage(data []byte) bool {
	var env Envelope
	if err := json.Unmarshal(data, &env); err != nil {
		return false
	}
	return env.Type != ""
}

File: packages/protocol/src/messages.ts
// ---- Text frame messages (sent as JSON strings over WebSocket) ----

/** Tunnel configuration sent from CLI to server during auth. */
export interface TunnelConfig {
  /** IP addresses or CIDR ranges allowed to access the tunnel. */
  allowedIps?: string[];
  /** Max requests per minute per source IP (0 = unlimited). */
  rateLimit?: number;
  /** Enable permissive CORS headers on all responses. */
  cors?: boolean;
  /** Custom response headers to inject. */
  customHeaders?: Record<string, string>;
}

export interface AuthMessage {
  type: "auth";
  subdomain: string;
  ttl?: number;
  sessionId?: string;
  /** Optional tunnel access-control and response configuration. */
  config?: TunnelConfig;
}

export interface AuthAckMessage {
  type: "auth-ack";
  subdomain: string;
  url: string;
  ttl: number;
  /** Actual seconds remaining on the alarm (may be less than ttl on resume) */
  remainingTtl: number;
  sessionId: string;
  maxBodySizeBytes: number;
  /** Echoed-back tunnel config accepted by the server. */
  config?: TunnelConfig;
}

export interface HttpRequestMessage {
  type: "http-request";
  id: string;
  method: string;
  path: string;
  headers: Record<string, string>;
  hasBody: boolean;
}

export interface HttpResponseMetaMessage {
  type: "http-response-meta";
  id: string;
  status: number;
  headers: Record<string, string>;
  hasBody: boolean;
}

export interface HttpBodyChunkMessage {
  type: "http-body-chunk";
  id: string;
  done: boolean;
}

export interface HttpRequestEndMessage {
  type: "http-request-end";
  id: string;
}

export interface HttpResponseEndMessage {
  type: "http-response-end";
  id: string;
}

export interface PingMessage {
  type: "ping";
}

export interface PongMessage {
  type: "pong";
}

export interface ErrorMessage {
  type: "error";
  message: string;
  requestId?: string;
  status?: number;
}

// ---- WebSocket relay messages (for proxying browser WS through tunnel) ----

export interface WsUpgradeMessage {
  type: "ws-upgrade";
  streamId: string;
  path: string;
  headers: Record<string, string>;
}

export interface WsUpgradeAckMessage {
  type: "ws-upgrade-ack";
  streamId: string;
  ok: boolean;
  error?: string;
}

export interface WsFrameMessage {
  type: "ws-frame";
  streamId: string;
  /** "text" or "binary" */
  frameType: "text" | "binary";
}

export interface WsCloseMessage {
  type: "ws-close";
  streamId: string;
  code: number;
  reason: string;
}

export type TunnelMessage =
  | AuthMessage
  | AuthAckMessage
  | HttpRequestMessage
  | HttpResponseMetaMessage
  | HttpBodyChunkMessage
  | HttpRequestEndMessage
  | HttpResponseEndMessage
  | PingMessage
  | PongMessage
  | ErrorMessage
  | WsUpgradeMessage
  | WsUpgradeAckMessage
  | WsFrameMessage
  | WsCloseMessage;

export function isTunnelMessage(data: unknown): data is TunnelMessage {
  return (
    typeof data === "object" &&
    data !== null &&
    "type" in data &&
    typeof (data as TunnelMessage).type === "string"
  );
}

export function parseTextMessage(raw: string): TunnelMessage | null {
  try {
    const parsed = JSON.parse(raw);
    if (isTunnelMessage(parsed)) return parsed;
    return null;
  } catch {
    return null;
  }
}
